GPU/GPUbenchmark/cuda/Kernels.cu

```diff
@@ -0,0 +1,288 @@
+    printDeviceProp(i);
+  }
+}
+
+template <class buffer_type>
+void GPUbenchmark<buffer_type>::generalInit(const int deviceId)
+{
+  cudaDeviceProp props;
+  size_t free;
+
+  // Fetch and store features
+  GPUCHECK(cudaGetDeviceProperties(&props, deviceId));
+  GPUCHECK(cudaMemGetInfo(&free, &mState.totalMemory));
+
+  mState.nMultiprocessors = props.multiProcessorCount;
+  mState.nMaxThreadsPerBlock = props.maxThreadsPerMultiProcessor;
+  mState.nMaxThreadsPerDimension = props.maxThreadsDim[0];
+  mState.scratchSize = static_cast<long int>(FREE_MEMORY_FRACTION_TO_ALLOCATE * free);
+  std::cout << ">>> Running on: " << props.name << std::endl;
+  // Allocate scratch on GPU
+  GPUCHECK(cudaMalloc(reinterpret_cast<void**>(&mState.scratchPtr), mState.scratchSize));
+  mState.computeScratchPtrs();
+
+  // Initialize corresponding buffers on host and copy content on GPU
+  mState.getHostBuffers().resize(mState.getScratchPtrs().size());
+  for (size_t iScratchPart{0}; iScratchPart < mState.getScratchPtrs().size(); ++iScratchPart) {
+    mState.getHostBuffers()[iScratchPart].resize(gpuState<buffer_type>::getArraySize());
+    GPUCHECK(cudaMemcpy(mState.getScratchPtrs()[iScratchPart], mState.getHostBuffers()[iScratchPart].data(), gpuState<buffer_type>::getArraySize() * sizeof(buffer_type), cudaMemcpyHostToDevice));
+  }
+  std::cout << "    ├ Allocated: " << std::setprecision(2) << bytesToGB(mState.scratchSize) << "/" << std::setprecision(2) << bytesToGB(mState.totalMemory)
+            << "(GB) [" << std::setprecision(3) << (100.f) * (mState.scratchSize / (float)mState.totalMemory) << "%]\n"
+            << "    ├ Number of scratch partitions: " << mState.getMaxSegments() << " of " << PARTITION_SIZE_GB << "GB each\n"
+            << "    ├ Size of arrays in segments: " << gpuState<buffer_type>::getArraySize() << " elements" << std::endl
+            << "    └ Memory buffers copied from host to device"
+            << std::endl;
+}
+
+template <class buffer_type>
+void GPUbenchmark<buffer_type>::readingInit()
+{
+  mState.deviceReadingResultsPtrs.resize(mState.getMaxSegments());
+  mState.hostReadingResultsVector.resize(mState.getMaxSegments());
+  for (size_t iScratchPart{0}; iScratchPart < mState.getMaxSegments(); ++iScratchPart) {
+    GPUCHECK(cudaMalloc(reinterpret_cast<void**>(&(mState.deviceReadingResultsPtrs[iScratchPart])), sizeof(buffer_type)));
+  }
+}
+
+template <class buffer_type>
+void GPUbenchmark<buffer_type>::readingBenchmark()
+{
+  dim3 nBlocks{static_cast<uint32_t>(mState.nMultiprocessors / mState.getMaxSegments())};
+  dim3 nThreads{static_cast<uint32_t>(std::min(mState.nMaxThreadsPerDimension, mState.nMaxThreadsPerBlock))};
+
+  for (size_t iScratchPart{0}; iScratchPart < mState.getMaxSegments(); ++iScratchPart) {
+    gpu::readerKernel<buffer_type><<<nBlocks, nThreads>>>(1, mState.deviceReadingResultsPtrs[iScratchPart], mState.getScratchPtrs()[iScratchPart], gpuState<buffer_type>::getArraySize());
```

Comment:
If I understand correctly, the different scratch parts should run in parallel. If you want to execute them in parallel on the GPU, you have to create 1 cudaStream per part, and run them in parallel.
Alternatively, and probably even better, you could select the scratchPart based on the `blockIdx.x` and start as many blocks as `mState.getMaxSegments()`, or a multipel thereof and then use multiple blocks per part.

Reply:
Yes, this was just a small shortcut to get something to  crunch, I was indeed planning something more sophisticated.

Reply:
I opted for the second idea, the pointer to scratch is passed to the kernel and then it computes the partition where to read based on `blockIdx.x`.